### ISSUE TITLE
embedded artwork not retreived for internet .mp3 streams

### ISSUE THREAD

**Issue by whatdoineed2do** [ejurgensen/forked-daapd#904]
This [internet mp3 /podcast has an embedded image](http://open.live.bbc.co.uk/mediaselector/6/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p083d4jg.mp3) but it's not extracted by the `artwork` subsystem.

The server performs:
```
[DEBUG]  artwork: Artwork request for item 321
[ SPAM]  artwork: Checking group source 'cache'
[ SPAM]  artwork: Checking group source 'directory'
[ SPAM]  artwork: Checking item source 'stream'
[ SPAM]  artwork: Trying internet stream artwork in http://open.live.bbc.co.uk/mediaselector/6/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p083d4jg.mp3
[ SPAM]  artwork: Checking item source 'playlist own'
[ SPAM]  artwork: Trying own artwork file https://podcasts.files.bbci.co.uk/p02pc9pj.jpg
[ SPAM]  artwork: Trying own artwork file https://podcasts.files.bbci.co.uk/p02pc9pj.png
[DEBUG]  artwork: No artwork found for item 321
```
What would be the best way to extract/provide artwork for podcasts like this?

**Comment by ejurgensen** [comment]
It might be enough to just add "| (1 << DATA_KIND_HTTP)" here:
https://github.com/ejurgensen/forked-daapd/blob/master/src/artwork.c#L185

**Comment by whatdoineed2do** [comment]
Magic, yes that makes it work for the web interface.  On IOS remote it only works when we set `artwork_individual = true`.

Leave you to commit?

```[DEBUG]  artwork: Artwork request for item 321
[ SPAM]  artwork: Checking group source 'cache'
[ SPAM]  artwork: Checking group source 'directory'
[ SPAM]  artwork: Checking item source 'embedded'
[ SPAM]  artwork: Trying embedded artwork in http://open.live.bbc.co.uk/mediaselector/6/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p083d4jg.mp3
[ SPAM]  artwork: Getting artwork (max destination width 600 height 600)
[DEBUG]  artwork: Original image dimensions: w 800 h 800
[DEBUG]  artwork: Rescale required, destination width 600 height 600
[DEBUG]  artwork: Artwork for 'Everybody's Shuffling!' found in source 'embedded'
```

**Comment by ejurgensen** [comment]
I tried the podcast address above to find out why Remote does not show artwork. However, I don't get any artwork anywhere, and the few versions of ffprobe I tried also tell me that the mp3 only has audio. What does your ffprobe say?

Without being able to reproduce I am hesitant about committing a change.

**Comment by whatdoineed2do** [comment]
I've got ffmpeg 4.0.3 on this dex box (fedora) and it can find a cover artwork.  I've got my test RSS branch with this on my RPI and its also working with ffmpeg 4.1.4 
```
$ ffprobe -i http://open.live.bbc.co.uk/mediaselector/6/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p083d4jg.mp3
ffprobe version 4.0.3 Copyright (c) 2007-2018 the FFmpeg developers
  built with gcc 8 (GCC)
  configuration: --prefix=/usr --bindir=/usr/bin --datadir=/usr/share/ffmpeg --docdir=/usr/share/doc/ffmpeg --incdir=/usr/include/ffmpeg --libdir=/usr/lib64 --mandir=/usr/share/man --arch=x86_64 --optflags='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection' --extra-ldflags='-Wl,-z,relro -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld ' --extra-cflags=' ' --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libvo-amrwbenc --enable-version3 --enable-bzlib --disable-crystalhd --enable-fontconfig --enable-frei0r --enable-gcrypt --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libcdio --enable-libdrm --enable-indev=jack --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libmp3lame --enable-nvenc --enable-openal --enable-opencl --enable-opengl --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-librsvg --enable-libsoxr --enable-libspeex --enable-libtheora --enable-libvorbis --enable-libv4l2 --enable-libvidstab --enable-libvpx --enable-libx264 --enable-libx265 --enable-libxvid --enable-libzvbi --enable-avfilter --enable-avresample --enable-postproc --enable-pthreads --disable-static --enable-shared --enable-gpl --disable-debug --disable-stripping --shlibdir=/usr/lib64 --enable-libmfx --enable-runtime-cpudetect
  libavutil      56. 14.100 / 56. 14.100
  libavcodec     58. 18.100 / 58. 18.100
  libavformat    58. 12.100 / 58. 12.100
  libavdevice    58.  3.100 / 58.  3.100
  libavfilter     7. 16.100 /  7. 16.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  1.100 /  5.  1.100
  libswresample   3.  1.100 /  3.  1.100
  libpostproc    55.  1.100 / 55.  1.100
Input #0, mp3, from 'http://open.live.bbc.co.uk/mediaselector/6/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p083d4jg.mp3':
  Metadata:
    title           : Everybody's Shuffling!
    artist          : BBC Radio 4
    album           : Friday Night Comedy from BBC Radio 4
    genre           : Podcast
    copyright       : (c) BBC 2020
    comment         : Nish is joined this week by Kerry Godliman, Fern Brady, Andy Zaltzman and Elis James.
    date            : 2020
  Duration: 00:28:58.61, start: 0.025056, bitrate: 128 kb/s
    Stream #0:0: Audio: mp3, 44100 Hz, stereo, fltp, 128 kb/s
    Metadata:
      encoder         : Lavc57.89
    Stream #0:1: Video: mjpeg, yuvj420p(pc, bt470bg/unknown/unknown), 800x800 [SAR 72:72 DAR 1:1], 90k tbr, 90k tbn, 90k tbc
    Metadata:
      comment         : Cover (front)
```
Running `mpv` on this stream also finds the artwork

**Comment by whatdoineed2do** [comment]
What I've just noticed is that if you simply add the stream to the server's queue (via the web i/f "add stream") then no artwork is searched.

I already this same stream in the library (as added by the RSS feed) as a non library item and when I add that item to the queue, the server goes and figures out that it needs to get the artwork and I see this:
```
[DEBUG]  artwork: Artwork request for item 321
[DEBUG]  artwork: Original image dimensions: w 800 h 800
[DEBUG]  artwork: Rescale required, destination width 600 height 600
[DEBUG]  artwork: Artwork for 'Everybody's Shuffling!' found in source 'embedded'
```
This I've tested with removing the `cache.db` and also without.

**Comment by whatdoineed2do** [comment]
One more observation:

* add internet stream (above) to `wtf.m3u` that is in the library
* add track to Q from web i/f
** artwork is pulled into cache for web i/f
* goto  IOS remote app nowplaying
** no artwork shown
* clear Q
* goto IOS remote app
** locate playlist `wtf` playlist
** add track
** goto now playing
** artwork is shown

**Comment by ejurgensen** [comment]
Weird, here is what I get (Raspbian Buster with ffmpeg 4.1.4):
```
$ ffprobe -i http://open.live.bbc.co.uk/mediaselector/6/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p083d4jg.mp3
ffprobe version 4.1.4-1+rpt7~deb10u1 Copyright (c) 2007-2019 the FFmpeg developers
  built with gcc 8 (Raspbian 8.3.0-6+rpi1)
  libavutil      56. 22.100 / 56. 22.100
  libavcodec     58. 35.100 / 58. 35.100
  libavformat    58. 20.100 / 58. 20.100
  libavdevice    58.  5.100 / 58.  5.100
  libavfilter     7. 40.101 /  7. 40.101
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  3.100 /  5.  3.100
  libswresample   3.  3.100 /  3.  3.100
  libpostproc    55.  3.100 / 55.  3.100
[mp3 @ 0x1236a00] Estimating duration from bitrate, this may be inaccurate
Input #0, mp3, from 'http://open.live.bbc.co.uk/mediaselector/6/redir/version/2.0/mediaset/audio-nondrm-download/proto/http/vpid/p083d4jg.mp3':
  Metadata:
    title           : Everybody's Shuffling!
    artist          : BBC Radio 4
    album           : Friday Night Comedy from BBC Radio 4
    genre           : Podcast
    copyright       : (c) BBC 2020
    date            : 2020-20-20
  Duration: 00:29:29.45, start: 0.000000, bitrate: 128 kb/s
    Stream #0:0: Audio: mp3, 44100 Hz, stereo, fltp, 128 kb/s

```

The artwork module doesn't handle non-library items, but it should work when you add it via m3u. Since I can't get the above link to work, I will try to upload a mp3 somewhere and see if I can reproduce with that.

I am also thinking about how to make the artwork module work with non-library items.

**Comment by chme** [comment]
@ejurgensen I have a branch that tries to solve the non-library artwork (I think this branch is it: https://github.com/chme/forked-daapd/commits/artwork_refactor). It comes with a bigger refactoring. Basically the artwork handlers will have to be refactored and a new type has to be implemented that fetches artwork for a queue item. I'll try to open PRs with smallish changes to keep them reviewable (and minimize breakage).

**Comment by ejurgensen** [comment]
Sounds good @chme, I will look forward to that. I have now tested with an mp3 I uploaded, and which I added to the library via a m3u. Artwork showed up as it should in Remote. Only when playing it as a non-library item does artwork not show up. So now I have added support for http sources to the embedded handler, but by accident I included it in commit 248f82f.
